folding-schemes/src/folding/nova/decider.rs

```diff
@@ -0,0 +1,496 @@
+
+        let zero = (&C2::BaseField::zero(), &C2::BaseField::zero());
+        let cmE_affine = proof.cf_U_i.cmE.into_affine();
+        let cmW_affine = proof.cf_U_i.cmW.into_affine();
+        let (cf_cmE_x, cf_cmE_y) = cmE_affine.xy().unwrap_or(zero);
+        let cf_cmE_z = C1::ScalarField::one();
+        let (cf_cmW_x, cf_cmW_y) = cmW_affine.xy().unwrap_or(zero);
+        let cf_cmW_z = C1::ScalarField::one();
+
+        // snark proof 1
+        let c1_public_input: Vec<C1::ScalarField> = [
+            vec![vp.pp_hash, i],
+            z_0,
+            z_i,
+            // U_{i+1} values:
+            vec![U.u],
+            U.x.clone(),
+            cmE_x,
+            cmE_y,
+            cmW_x,
+            cmW_y,
+            // CS1 values:
+            proof.cs1_challenges.to_vec(), // c_W, c_E
+            vec![
+                proof.cs1_proofs[0].eval, // eval_W
+                proof.cs1_proofs[1].eval, // eval_E
+            ],
+            // cf_U_i values
+            NonNativeUintVar::<CF2<C2>>::inputize(proof.cf_U_i.u),
+            proof
+                .cf_U_i
+                .x
+                .iter()
+                .flat_map(|&x_i| NonNativeUintVar::<CF2<C2>>::inputize(x_i))
+                .collect::<Vec<C1::ScalarField>>(),
+            vec![
+                *cf_cmE_x, *cf_cmE_y, cf_cmE_z, *cf_cmW_x, *cf_cmW_y, cf_cmW_z,
+            ],
+            // NIFS values:
+            cmT_x,
+            cmT_y,
+            vec![proof.r],
+        ]
+        .concat();
+
+        let c1_snark_v = S1::verify(&vp.c1_snark_vp, &c1_public_input, &proof.c1_snark_proof)
+            .map_err(|e| Error::Other(e.to_string()))?;
+        if !c1_snark_v {
+            return Err(Error::SNARKVerificationFail);
+        }
+        dbg!("snark1 verified");
+
+        let (cf2_cmE_x, cf2_cmE_y) = NonNativeAffineVar::inputize(proof.cf_U_i.cmE)?;
+        let (cf2_cmW_x, cf2_cmW_y) = NonNativeAffineVar::inputize(proof.cf_U_i.cmW)?;
+
+        // snark proof 2
+        // migrate pp_hash from C1::Fr to C1::Fq
+        let pp_hash_Fq =
+            C2::ScalarField::from_le_bytes_mod_order(&vp.pp_hash.into_bigint().to_bytes_le());
+        let c2_public_input: Vec<C2::ScalarField> = [
+            vec![pp_hash_Fq],
+            vec![proof.cf_U_i.u],
+            proof.cf_U_i.x.clone(),
+            cf2_cmE_x,
+            cf2_cmE_y,
+            cf2_cmW_x,
+            cf2_cmW_y,
+            proof.cs2_challenges.to_vec(),
+            vec![
+                proof.cs2_proofs[0].eval, // eval_W
+                proof.cs2_proofs[1].eval, // eval_E
+            ],
+        ]
+        .concat();
+
+        let c2_snark_v = S2::verify(&vp.c2_snark_vp, &c2_public_input, &proof.c2_snark_proof)
+            .map_err(|e| Error::Other(e.to_string()))?;
+        if !c2_snark_v {
+            return Err(Error::SNARKVerificationFail);
+        }
+        dbg!("snark2 verified");
```

Comment:
Do you want to keep all the `dbg!`s in the tests? I think it can pollute a bit the review of all the test run. But whatever. Not strongly positioned.

Reply:
removed the `dbg`s https://github.com/privacy-scaling-explorations/sonobe/compare/04bef2887e8b26786a0c15c1f68826394f1163a0..c07088ef25c4daadd94820c2a906d0d9803cd7ec